packages/kbn-eslint-plugin-eslint/rules/no_restricted_paths.js

```diff
@@ -0,0 +1,136 @@
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+const path = require('path');
+const resolve = require('eslint-module-utils/resolve').default;
+const mm = require('micromatch');
+
+function isStaticRequire(node) {
+  return (
+    node &&
+    node.callee &&
+    node.callee.type === 'Identifier' &&
+    node.callee.name === 'require' &&
+    node.arguments.length === 1 &&
+    node.arguments[0].type === 'Literal' &&
+    typeof node.arguments[0].value === 'string'
+  );
+}
+
+function resolveBasePath(basePath) {
+  return path.isAbsolute(basePath) ? basePath : path.relative(process.cwd(), basePath);
+}
+
+function traverseToTopFolder(src, pattern) {
+  while (mm([src], pattern).length > 0) {
+    const srcIdx = src.lastIndexOf(path.sep);
+    src = src.slice(0, srcIdx);
+  }
+  return src;
+}
+
+function isSameFolderOrDescendent(src, imported, pattern) {
+  const srcFileFolderRoot = traverseToTopFolder(src, pattern);
+  const importedFileFolderRoot = traverseToTopFolder(imported, pattern);
+  return srcFileFolderRoot === importedFileFolderRoot;
+}
+
+module.exports = {
+  meta: {
+    schema: [
+      {
+        type: 'object',
+        properties: {
+          zones: {
+            type: 'array',
+            minItems: 1,
+            items: {
+              type: 'object',
+              properties: {
+                target: {
+                  anyOf: [{ type: 'string' }, { type: 'array', items: { type: 'string' } }],
+                },
+                from: {
+                  anyOf: [{ type: 'string' }, { type: 'array', items: { type: 'string' } }],
+                },
+                allowSameFolder: { type: 'boolean' },
+              },
+              additionalProperties: false,
+            },
+          },
+          basePath: { type: 'string' },
+        },
+        additionalProperties: false,
+      },
+    ],
+  },
+
+  create(context) {
+    const options = context.options[0] || {};
+    const zones = options.zones || [];
+    const basePath = options.basePath ? resolveBasePath(options.basePath) : process.cwd();
```

Comment:
Why not just require `options.basePath`? I think it would make the behavior of this rule more explicit and easier to reason about.

```js
const basePath = options.basePath;
if (!basePath || !isAboslute(basePath)) {
  throw new Error('basePath option must be specified and must be absolute')
}
```